pkg/morph/deploy/notary.go

```diff
@@ -869,3 +878,353 @@ func makeUnsignedDesignateCommitteeNotaryTx(roleContract *rolemgmt.Contract, com
+				// using worker pool, but actions seem to be relatively lightweight
+
+				const expectedSigners = 3 // sender + committee + Notary
+				mainTx := notaryEvent.NotaryRequest.MainTransaction
+				// note: instruction above can throw NPE and it's ok to panic: we confidently
+				// expect that only non-nil pointers will come from the channel (NeoGo
+				// guarantees)
+
+				srcMainTxHash := mainTx.Hash()
+				_, processed := mProcessedMainTxs[srcMainTxHash]
+
+				switch {
+				case processed:
+					prm.logger.Info("main transaction of the notary request has already been processed, skip",
+						zap.Stringer("ID", srcMainTxHash))
+					continue
+				case notaryEvent.Type != mempoolevent.TransactionAdded:
+					prm.logger.Info("unsupported type of the notary request event, skip",
+						zap.Stringer("got", notaryEvent.Type), zap.Stringer("expect", mempoolevent.TransactionAdded))
+					continue
+				case len(mainTx.Signers) != expectedSigners:
+					prm.logger.Warn("unsupported number of signers of main transaction from the received notary request, skip",
+						zap.Int("expected", expectedSigners), zap.Int("got", len(mainTx.Signers)))
+					continue
+				case !mainTx.HasSigner(committeeMultiSigAccID):
+					// in theory, there can be any notary requests besides those sent by the current
+					// auto-deploy procedure. However, it's better to log with 'info' severity since
+					// this isn't really expected in practice.
+					prm.logger.Warn("committee is not a signer of main transaction from the received notary request, skip")
+					continue
+				case mainTx.HasSigner(prm.localAcc.ScriptHash()):
+					prm.logger.Info("main transaction from the received notary request is signed by a local account, skip")
+					continue
+				case len(mainTx.Scripts) == 0:
+					prm.logger.Warn("missing scripts of main transaction from the received notary request, skip")
+					continue
+				}
+
+				bSenderKey, ok := vm.ParseSignatureContract(mainTx.Scripts[0].VerificationScript)
+				if !ok {
+					prm.logger.Info("first verification script in main transaction of the received notary request is not a signature one, skip", zap.Error(err))
+					continue
+				}
+
+				senderKey, err := keys.NewPublicKeyFromBytes(bSenderKey, elliptic.P256())
+				if err != nil {
+					prm.logger.Info("failed to decode sender's public key from first script of main transaction from the received notary request, skip", zap.Error(err))
+					continue
+				}
+
+				// copy transaction to avoid pointer mutation. For simplicity, make a shallow
+				// copy: we change only Scripts field below. Overall, this approach is very
+				// risky, and it'd better to make a deep copy
+				mainTxCp := *mainTx
+				mainTxCp.Scripts = nil
+
+				mainTx = &mainTxCp // source one isn't needed anymore
+
+				// it'd be safer to get into the transaction and analyze what it is trying to do.
+				// For simplicity, now we blindly sign it
```

Comment:
Either fix this in the current PR or create an issue for that.

Reply:
created #2430